src/man/man3/sam_audit.3
.\" $Revision: 1.17 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.TH sam_audit 3 "07 Jan 2009"
.SH NAME
sam_audit \- Audits media
.SH SYNOPSIS
.LP
.BI "cc [ " "flag"
.BI " ... ] " "file"
.BI " ... -L/opt/SUNWsamfs/lib -lsamapi [ " "library" " ... ]"
.LP
.nf
.ft 3
#include "/opt/SUNWsamfs/include/samapi.h"
.ft
.fi
.LP
.BI "int sam_audit(ushort_t " "eq_number" ,
.BI "uint_t " "ea" ,
.BI "int " "wait_response" );
.SH DESCRIPTION
.PP
.B sam_audit(\|)
performs a catalog audit on either a single element address in a robot or all the element addresses
in a robot, for either optical or tape robotic devices.  Specifying 
.I ea
as the number of the robot element address will perform a catalog audit of that element address.
If the 
.I ea
is set to the global
.B ROBOT_NO_SLOT
value, an audit of the entire robot will be performed. The call will return
immediately after issuing the command if zero is specified for
.I wait_response
value.  Other values for
.I wait_response
will give undefined results.
.SH "RETURN VALUES"
Upon succesful completion a value of 0 is returned.
Otherwise, a value of \-1 is returned and
\f4errno\fP
is set to indicate the error.
.SH ERRORS
.PP
.B sam_audit(\|)
fails if one or more of the following are true:
.TP 20
.SB ER_FIFO_PATH_LENGTH
The path and filename for the FIFO pipe is too long.
.TP
.SB ER_NO_EQUIP_ORDINAL
.I eq_number
is not a defined equipment number in this configuration.
.TP
.SB ER_NO_MASTER_SHM
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment defined.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NO_MASTER_SHM_ATT
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment found.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NOT_VALID_SLOT_NUMBER
.I ea
is not a valid element address in the robot at equipment number
.I eq_number
.TP
.SB ER_NO_RESPONSE_FIFO
Unable to create the response FIFO pipe.
.TP
.SB ER_NO_STAT_ROBOT_CATALOG
The robot media changer catalog for equipment number
.I eq_number
cannot be accessed for status.
.TP
.SB ER_OPERATOR_NOT_PRIV
Operator does not have permission to perform a full audit
.TP
.SB ER_ROBOT_CATALOG_MISSING
The robot media changer catalog for equipment number
.I eq_number
is missing and a full audit is required.
.TP
.SB ER_SLOT_NOT_OCCUPIED
No media exists at the element address number
.I ea
specified
.TP
.SB ER_UNABLE_TO_MAP_CATALOG
The catalog for the removable media changer at equipment number
.I eq_number
is unable to be mapped into memory.
.SH FILES
.TP 20
.SB mcf
The configuration file for Sun StorEdge \%SAM-FS or Sun \%SAM-QFS
.SH "SEE ALSO"
.BR auditslot (8),
.BR export (8),
.BR import (8),
.BR move (8),
.BR sam-robotsd (8).
.PP
.BR mcf (5).


src/lib/samapi/sam_audit.c
/*
 * sam_audit.c - audit removable media API
 *
 *	sam_audit() causes the device specified to be audited
 *
 */


/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.22 $"

static char *_SrcFile = __FILE__; /* Using __FILE__ makes duplicate strings */

#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/mman.h>
#include <sys/stat.h>

#include "sam/types.h"
#include "sam/defaults.h"
#include "aml/device.h"
#include "aml/fifo.h"
#include "sam/param.h"
#include "aml/proto.h"
#include "aml/sam_utils.h"
#include "aml/catalog.h"
#include "aml/catlib.h"
#include "aml/samapi.h"
#include "pub/sam_errno.h"


/*
 *	sam_audit() - API function to audit media, including entire robot
 *
 *	Input parameters --
 *		eq_ord		Equipment ordinal
 *		ea		Element address in robot to be audited or
 *				ROBOT_NO_SLOT
 *		wait_response	Wait for command response if nonzero
 *				-1	Wait forever
 *				> zero	Wait this many seconds for response
 *
 *	Output parameters --
 *		errno		Set to error number if error encountered
 *
 *	Return value --
 *		0		If successful in auditing the media
 *		-1		If error encountered,
 *				'errno' set to error number
 *
 */

int
sam_audit(
	ushort_t eq_ord,	/*	Equipment ordinal */
	uint_t	 ea,		/*	Element address */
	int wait_response)	/*	Nonzero value to wait for response */
{
	char *fifo_path;	/*	Path to FIFO pipe, from shared memory */
	dev_ent_t *device;
	operator_t operator;	/*	Data on operator */
	sam_cmd_fifo_t	cmd_block;
	struct CatalogEntry ced;
	struct CatalogEntry *ce = &ced;

	if (CatalogInit("SamApi") == -1) {
		errno = ER_UNABLE_TO_INIT_CATALOG;
		/* Unable to initialize catalog. */
		return (-1);
	}

	/*
	 * Get device entry and FIFO path from master shared memory segment
	 */

	if (sam_get_dev(eq_ord, &device, &fifo_path, &operator) < 0) {
		return (-1);
	}

	if ((!IS_ROBOT(device)) || (device->state > DEV_IDLE)) {
		errno = ER_NO_AUDIT;	/* "Cannot audit device" */
		return (-1);
	}

	/*
	 * Clear command block
	 */

	memset(&cmd_block, 0, sizeof (cmd_block));


	/*
	 *	If a ea is specified, verify that it is valid
	 */

	if ((uint_t)ROBOT_NO_SLOT != ea) {
		ce = CatalogGetCeByLoc(eq_ord, ea, 0, &ced);
		if (ce == NULL) {
			errno = ER_NOT_VALID_SLOT_NUMBER;
			return (-1);
		}
		if (!(ce->CeStatus & CES_inuse)) {
			errno = ER_SLOT_NOT_OCCUPIED; /* Slot is not occupied */
			return (-1);
		}
	}
	else
	{
		/*
		 *	Check if operator permitted to do a full audit
		 */

		if (!SAM_OPER_FULLAUDIT(operator)) {
			errno = ER_OPERATOR_NOT_PRIV;
			return (-1);
		}
	}

	/*
	 *	Put command block values into command
	 */

	cmd_block.magic = CMD_FIFO_MAGIC;
	cmd_block.cmd = CMD_FIFO_AUDIT;
	cmd_block.slot = ea;
	cmd_block.eq = eq_ord;
	cmd_block.part = 0;

	/*
	 *	Send command and wait for response if requested
	 */

	if (sam_send_cmd(&cmd_block, wait_response, fifo_path) < 0) {
		return (-1);
	}

	return (0);
}
